validation/helpers.py

```diff
@@ -0,0 +1,195 @@
+                    continue
+
+        # no change, carry on
+        if "+" not in c and "-" not in c:
+            continue
+
+        # added or removed a hyphen
+        if c == "- -" or c == "+ -":
+            continue
+
+        # check for swapping glides
+        if c == "- y":
+            if i + 1 < len(df):
+                if df[i + 1] == "+ w":
+                    continue
+
+        if c == "- w":
+            if i + 1 < len(df):
+                if df[i + 1] == "+ y":
+                    continue
+
+        if c == "+ y":
+            if i - 1 >= 0:
+                if df[i - 1] == "- w":
+                    continue
+
+        if c == "+ w":
+            if i - 1 >= 0:
+                if df[i - 1] == "- y":
+                    continue
+
+        # end check for swapping glides
+
+        if c.split()[1] in vowels:
+            # adding or removing a diacritic from a vowel has cost 0
+            alt = get_analog(c)
+            if i + 1 < len(df):
+                if df[i + 1] == alt:
+                    continue
+            if i - 1 >= 0:
+                if df[i - 1] == alt:
+                    continue
+
+        if "i" in c:
+            # adding an -i- between two consonants has cost 0.5
+            j = i + 1 if i + 1 < len(df) else None
+            k = i - 1 if i - 1 >= 0 else None
+            if j and k:
+                if df[j] in consonants and df[k] in consonants:
+                    med += 0.5
+                    continue
+
+        if "h" in c:
+            # did we just add an -h- between a vowel and a consonant?
+            j = i + 1 if i + 1 < len(df) else None
+            k = i - 1 if i - 1 >= 0 else None
+            if j and k:
+                after = df[j].replace("-", "")
+                after = after.replace("+", "")
+                before = df[k].replace("-", "")
+                before = before.replace("+", "")
+                after = "  " + after.strip()
+                before = before.strip()
+
+                if after in consonants and before in vowels:
+                    # we did, this costs nothing
+                    continue
+
+        # everything else has cost 1
+        med += 1
+
+    return med
+
+
+def get_analog(char):
+    # take in '- vowel'
+    # return '+ vowel with hat'
+    # OR
+    # take in '+ vowel'
+    # return '- vowel with hat'
+
+    alt = ""
+    if "-" in char:
+        alt = char.replace("-", "+")
+    else:
+        alt = char.replace("+", "-")
+
+    if char.split()[1] in ["a", "i", "o"]:
+        alt = alt.replace("a", "â")
+        alt = alt.replace("i", "î")
+        alt = alt.replace("o", "ô")
+
+    elif char.split()[1] in ["â", "î", "ô"]:
+        alt = alt.replace("â", "a")
+        alt = alt.replace("î", "i")
+        alt = alt.replace("ô", "o")
+
+    return alt
+
+
+def get_translations_from_itwewina(word):
+    url = "https://sapir.artsrn.ualberta.ca/cree-dictionary/click-in-text/?q=" + word
+    r = requests.get(url, verify=False)
```

Comment:
```suggestion
    r = requests.get(url)
```

Sapir was misconfigured and I updated it. If you still have issues we might need the `certifi` package in the Pipenv. The quick hack fix was to paste the correct intermediate certificate onto sapir. cc: @eddieantonio 

If anyone is interested, I can explain this all in great depth, but am going to assume other people are happy they don’t have to think about it.

```
--- /etc/apache2/sites-available/cree-dictionary-python-3.9.orig	2021-01-29 10:24:01.059058416 -0700
+++ /etc/apache2/sites-available/cree-dictionary-python-3.9.conf	2021-01-29 10:22:02.592036453 -0700
@@ -57,7 +57,7 @@
 #   the referenced file can be the same as SSLCertificateFile
 #   when the CA certificates are directly appended to the server
 #   certificate for convinience.
-SSLCertificateChainFile /etc/ssl/certs/intermediate.pem
+SSLCertificateChainFile /etc/ssl/certs/GlobalSign_RSA_OV_SSL_CA_2018.pem
 
 #   Certificate Authority (CA):
 #   Set the CA certificate verification path where to find CA
```

Reply:
Nice! I had been meaning to make this an issue for itwêwina, thanks for jumping on it!

Reply:
@andrewdotn thanks for fixing this!

I don't really understand what was misconfigured or why, so I'd appreciate an explanation, but that can be whenever :)

Reply:
Run `openssl s_client -connect sapir.artsrn.ualberta.ca:443 -showcerts < /dev/null` to see a dump of the certs being served up. (*Aside*: an advanced TLS server, such as the one in the go standard library, can dynamically serve up different certificates depending on the capabilities of the client. So this doesn’t work 100% of the time for debugging, but is fine for the predominant case where the server has a static config.)

To see details of individual certs, copy the `-----BEGIN CERTIFICATE-----…-----END CERTIFICATE-----` bit and do `getclip | openssl x509 -text -noout`

The sapir.artsrn.ualberta.ca certificate has `Issuer: C=BE, O=GlobalSign nv-sa, CN=GlobalSign RSA OV SSL CA 2018`, but the intermediate that was being sent (from `intermediate.pem`) had `Subject: C=BE, O=GlobalSign nv-sa, CN=GlobalSign Organization Validation CA - SHA256 - G2`.

It’s likely—but I haven’t checked—that your browser already comes with that specific intermediate certificate, but Python doesn’t, which is why the certificate validates in your browser but wasn’t validating from Python’s `requests.get()`.

The quick and dirty way to find an intermediate certificate is to search the web for sapir’s certificate’s “X509v3 Authority Key Identifier: keyid”, [F8:EF:7F:F2:CD:78:67:A8:DE:6F:8F:24:8D:88:F1:87:03:02:B3:EB](https://google.ca/search?q=F8:EF:7F:F2:CD:78:67:A8:DE:6F:8F:24:8D:88:F1:87:03:02:B3:EB), then look for a PEM Download link on one of the result pages, paste it into a file on the webserver, `apache2ctl -t && apache2ctl restart`, then see if that fixed the problem.

If that fixes things, then that’s usually good enough. There can be cases where it gets really confusing because certificate-authority certificates can be reissued with the same public key but different names/expiration dates/signature algorithms/serial numbers/&c., and some of those variants may be trusted by some clients but not others, which is part of why go supports dynamic certificate selection; but fortunately I’ve never had to deal with that myself.